jellyfin/jellyfin-androidtv
# Network State Handling and Fix Login Error With No Network

**Changes**

- Add a Util `NetworkStatus` to register for [NetworkCallbacks](https://developer.android.com/reference/kotlin/android/net/ConnectivityManager.NetworkCallback)
- Make this observable / stateful (for toast and login flow)
- Add a toast tied to the application lifetime, that will notify users when the network disconnects / reconnects. This should allow users to actively see they're going to have problems. Especially, as the connection status on Android TV requires users to actively go to their settings, instead of glancing at a status bar.
- Add a check to `login` that the network is connected, and bail out if we're not to emit the correct error.

Fixes: #1241 

**To Test**
(I'd recommend using a virtual device for this, these are the tests I've run through locally too)
- Disconnect / Reconnect on the network whilst remaining in the app.
- If you are using a virtual device, ADB can be used to avoid going out / in the app: 
```
adb devices  # Check the device appears first
adb shell svc wifi disable
adb shell svc wifi enable
```
- Alternatively, yank the power cable out of the wireless AP / remove ethernet cable on a real device
- Check the toast appears for connecting / disconnecting
- Check the login screen gives a sensible prompt for starting the app without a connection
- Check you can sign in after re-enabling the connection / with a good connection from the start.
